src/service/gps-monitor.man
.\"
.TH gps-monitor 1 "Feb. 2012" "Ubuntu"
.SH NAME
gps-monitor \- Monitors GPS information on suitable UMTS/GPS/LTE modules
.SH SYNOPSIS
.B gps-monitor
.SH DESCRIPTION
.B gps-monitor
enables and queries the GPS information available with some modems. The 
information is stored at /var/run/ipcsystools/.
.SH AUTHOR
Carsten Schlote



ChangeLog
commit 9278db22cd4f9de47ed328ffcf40365fb2a190c5
Author: Carsten Schlote <c.schlote@konzeptpark.de>
Date:   Mon Nov 21 16:26:35 2011 +0100

    [ipsystools/umtscardtool] Commented out unused code
    
    - Removed unused local static function
    
    Signed-off-by: Carsten Schlote <c.schlote@konzeptpark.de>

commit 2b5fe34f8fbeb31f74ce5bdbc970d3562827cda1
Author: Carsten Schlote <c.schlote@konzeptpark.de>
Date:   Mon Nov 21 16:25:50 2011 +0100

    [ipsystools/umtscardtool] Changed paths
    
    - Changed path /var/log... to /var/run/...
      GI#NCU-530
    
    Signed-off-by: Carsten Schlote <c.schlote@konzeptpark.de>

commit 03237c4a8d49c8fa4626d3b17900faa92a3bf094
Author: Carsten Schlote <c.schlote@konzeptpark.de>
Date:   Fri Nov 18 15:42:46 2011 +0100

    [ipcsystools] Updated missing file
    
    Signed-off-by: Carsten Schlote <c.schlote@konzeptpark.de>

commit 1a83cc58eed229fb36e027022cbb0f6f42510651
Author: Carsten Schlote <c.schlote@konzeptpark.de>
Date:   Fri Nov 18 15:37:00 2011 +0100

    [ipcsystools] Added execute flags
    
    Signed-off-by: Carsten Schlote <c.schlote@konzeptpark.de>

commit 0cb3a9f2d7ba025ba4c40157e19fe8c3e28fed6f
Author: Carsten Schlote <c.schlote@konzeptpark.de>
Date:   Fri Nov 18 15:30:57 2011 +0100

    [ipcsystools] Updated deploy script to fabricate upstream releases
    
    - Modified script to
      - Switch to upstream branch
      - Merge current master into upstream
      - Create an annotated signed tag for upstream release
      - Export the release to an orig tarball
      - Use pristine-tar commit to memorize the tarball
    - Script does some sanity check. You should use this script on a
      clean working copy. In doubt, execute command from script by hand.
    
    Signed-off-by: Carsten Schlote <c.schlote@konzeptpark.de>

commit 9cdcaab24a76e0c60f033beb57c803206e2f54b5
Author: Carsten Schlote <c.schlote@konzeptpark.de>
Date:   Fri Nov 18 14:46:59 2011 +0100

    [ipcsystools] Fixed script permissions for ipclib.inc
    
    Signed-off-by: Carsten Schlote <c.schlote@konzeptpark.de>

commit 58819a1713712054637bdd0031ab25e85bdffb17
Author: Carsten Schlote <c.schlote@konzeptpark.de>
Date:   Fri Nov 18 14:37:38 2011 +0100

    [ipcsystools] Use invoke-rc.d instead of directly call init.d script
    
    - Legacy from ptxdist
    - Using invoke-rc.d now to call script
    
    Signed-off-by: Carsten Schlote <c.schlote@konzeptpark.de>

commit a99023dacdb5bd268654b47d8ba3ff8d5e0519e0
Author: Carsten Schlote <c.schlote@konzeptpark.de>
Date:   Fri Nov 18 14:31:58 2011 +0100

    [ipcsystools] Fixed install macro
    
    Signed-off-by: Carsten Schlote <c.schlote@konzeptpark.de>

commit ccd377396e8b3f89cdadf3e8417834a1d62eb543
Author: Carsten Schlote <c.schlote@konzeptpark.de>
Date:   Fri Nov 18 14:22:00 2011 +0100

    [ipcsystools] Moved package udev rule to /lib/udev/rules.d
    
    - Changed according to udev policy guide for Ubuntu
    
    Signed-off-by: Carsten Schlote <c.schlote@konzeptpark.de>

commit fcc6264b5e44039d4771f48f7f316e5b76932462
Author: Carsten Schlote <c.schlote@konzeptpark.de>
Date:   Fri Nov 18 14:07:20 2011 +0100

    [ipcsystools] Install scripts with execute flag set
    
    Signed-off-by: Carsten Schlote <c.schlote@konzeptpark.de>

commit 2461eaa1015cf4dd4fa1bda594a2067185252e69
Author: Carsten Schlote <c.schlote@konzeptpark.de>
Date:   Fri Nov 18 13:57:03 2011 +0100

    [ipcsystools] Moved scripts from init.d to /usr/sbin
    
    - As upstart is used, I moved the scripts from init.d to some
      appropriate dir as sysv init is not used here.
    - Silenced lintian errors
    
    Signed-off-by: Carsten Schlote <c.schlote@konzeptpark.de>
    [ipcsystools] WIPWIPWIPWIP

commit 1a66550d95bd64133d58fbd025337d190455ed70
Author: Carsten Schlote <c.schlote@konzeptpark.de>
Date:   Thu Nov 17 17:13:32 2011 +0100

    [ipcsystools] Final fixes, no syncs with binary snapshot on mIPC
    
    - Only LED tool is missing, should be separate package because of
      header dependancies on cgoslx package.
    
    Signed-off-by: Carsten Schlote <c.schlote@konzeptpark.de>

commit 2812f36bf315d1d1b6da5279a654c0df9e82699b
Author: Carsten Schlote <c.schlote@konzeptpark.de>
Date:   Thu Nov 17 16:55:43 2011 +0100

    [ipcsystools/scripts] Replaced mcb by ipc
    
    Signed-off-by: Carsten Schlote <c.schlote@konzeptpark.de>

commit 82447efab157701536100455aee205a055b8250d
Author: Carsten Schlote <c.schlote@konzeptpark.de>
Date:   Thu Nov 17 16:54:23 2011 +0100

    [ipcsystools] Renamed library mcbtoolset to ipctoolset et al.
    
    Signed-off-by: Carsten Schlote <c.schlote@konzeptpark.de>

commit fc276fbbdf89e19a9024cf2246327a51bf2a7923
Author: Carsten Schlote <c.schlote@konzeptpark.de>
Date:   Thu Nov 17 16:40:05 2011 +0100

    [ipcsystools/scripts] Renamed files
    
    Signed-off-by: Carsten Schlote <c.schlote@konzeptpark.de>

commit 9f4b47a310a154615a358fc575df1ab7642dec35
Author: Carsten Schlote <c.schlote@konzeptpark.de>
Date:   Thu Nov 17 16:21:07 2011 +0100

    [ipcsystools] Fixed autoconf Makefiles for proper make distclean
    
    - Do not delete Makefile.in files! These are part of a dist
      release and required for ./configure
    
    Signed-off-by: Carsten Schlote <c.schlote@konzeptpark.de>

commit 8e2d9300e23fea608596c9ea70dfe89fe853588c
Author: Carsten Schlote <c.schlote@konzeptpark.de>
Date:   Thu Nov 17 16:14:41 2011 +0100

    [ipcsystools/etc] Updated Makefile.am with changes
    
    - Updated install command sequence for new names and locations
    - Commented out no longer used files.
    
    Signed-off-by: Carsten Schlote <c.schlote@konzeptpark.de>

commit 04df1839eae9db2d5ef6089dc90a306a47c7cf0a
Author: Carsten Schlote <c.schlote@konzeptpark.de>
Date:   Thu Nov 17 16:13:41 2011 +0100

    [ipcsystools/etc/udev] Merge dev snapshot changes into source
    
    - Added udev snippet to configure umts modem with static name
    
    Signed-off-by: Carsten Schlote <c.schlote@konzeptpark.de>


src/utils/scripts/gps-monitor.sh
#!/bin/bash

#
# GPS Function Library
#

PATH=/bin:/usr/bin:/sbin:/usr/sbin
. /usr/share/mcbsystools/mcblib.inc

DESC="gps-monitor[$$]"

GPS_STATUS_FILE=$MCB_STATUSFILE_DIR/gps_status
GPS_LOCATION_FILE=$MCB_STATUSFILE_DIR/gps_location
GPS_SATINFO_FILE=$MCB_STATUSFILE_DIR/gps_satinfo

GPS_MONITOR_PID_FILE=/var/lock/gps-monitor.pid
GPS_MONITOR_STATUS=/var/run/gps-monitor-status

GPS_ENABLE=`getmcboption monitor.gps.enable`
GPS_FIXPREC=`getmcboption monitor.gps.fix_prec`
GPS_TRKPREC=`getmcboption monitor.gps.trk_prec`
GPS_START_BACKEND=`getmcboption monitor.gps.start_backend`
GPS_BACKENDS=(`getmcboption monitor.gps.backends`)

# ----------------------------------------------------------------------
# Errorcodes
# ----------------------------------------------------------------------

# AT command error codes (!GPSEND, !GPSSTATUS, !GPSTRACK)

GPS_Errorcode=(
"Phone is offline"
"No service"
"No connection with PDE (Position Determining Entity)"
"No data available"
"Session Manager is busy"
"Reserved"
"Phone is GPS-locked"
"Connection failure with PDE"
"Session ended because of error condition"
"User ended the session"
"End key pressed from UI"
"Network session was ended"
"Timeout (for GPS search)"
"Conflicting request for session and level of privacy"
"Could not connect to the network"
"Error in fix"
"Reject from PDE"
"Reserved"
"Ending session due to E911 call"
"Server error"
"Reserved"
"Reserved"
"Unknown system error"
"Unsupported service"
"Subscription violation"
"Desired fix method failed"
"Reserved"
"No fix reported because no Tx confirmation was received"
"Network indicated normal end of session"
"No error specified by the network"
"No resources left on the network"
"Position server not available"
"Network reported an unsupported version of protocol"
)

# AT command error codes (!GPSFIX)

GPS_ErrorCodeFix=(
"No error"
"Invalid client ID"
"(MC8775V / 80 / 81 / 90 / 90V) Bad service parameter (MC8785V) Reserved"
"Bad session type parameter"
"(MC8775V / 80 / 81 / 90 / 90V) Incorrect privacy parameter (MC8785V) Reserved"
"(MC8775V / 80 / 81 / 90 / 90V) Incorrect download parameter (MC8785V) Reserved"
"(MC8775V / 80 / 81 / 90 / 90V) Incorrect network access parameter (MC8785V) Reserved"
"Incorrect operation parameter"
"Incorrect number of fixes parameter"
"Incorrect server information parameter"
"Error in timeout parameter"
"Error in QOS accuracy threshold parameter"
"No active session to terminate"
"(MC8775V / 80 / 81 / 90 / 90V) Session is active (MC8785V) Reserved"
"Session is busy"
"Phone is offline"
"Phone is CDMA locked"
"GPS is locked"
"Command is invalid in current state"
"Connection failure with PDE"
"PDSM command buffer unavailable to queue command"
"Search communication problem"
"(MC8775V / 80 / 81 / 90 / 90V) Temporary problem reporting position determination results (MC8785V) Reserved"
"(MC8775V / 80 / 81 / 90 / 90V) Error mode not supported (MC8785V) Reserved"
"Periodic NI in progress"
"(MC8775V / 80 / 81 / 90 / 90V) Unknown error (MC8785V) Client authentication failure"
"(MC8785V) Unknown error"
)

# ----------------------------------------------------------------------
# Modem Handling
# ----------------------------------------------------------------------

# at!custom?
# !CUSTOM:
#             PUKPRMPT            0x01
#             MEPCODE             0x01
#             ISVOICEN            0x01
#             PRLREGION           0x01
#             PCSCDISABLE         0x03
#             GPSENABLE           0x01  (or not displayed when not configured)
# OK
function gps_checkmodem_feature ()
{
    syslogger "debug" "Check modem caps"
    if ! $MCB_TOOLS_DIR/umtscardtool -s at!custom? | grep -q GPSENABLE; then
	syslogger "error" "Modem has no GPS feature support"
	return 1
    fi
    return 0
}

# at!gpsautostart=?
# !GPSAUTOSTART: <enable>[,<fixtype>,<maxtime>,<maxdist>,<fixrate>]
# <enable>:  0-Disabled, 1-Enabled
# <fixtype>: 1-Standalone, 2-MS-Based, 3-MS-Assisted
# <maxtime>: 1-255 seconds
# <maxdist>: 1-4294967280 meters
# <fixrate>: 1-65535 seconds
function gps_gpsautostart ()
{
    local ena=1; [ -n "$1" ] && mode=$1
    local mode=1; [ -n "$2" ] && mode=$2
    local to=15; [ -n "$3" ] && to=$3
    local dist=30; [ -n "$4" ] && dist=$4
    local fr=10; [ -n "$4" ] && dist=$4
    syslogger "debug" "Send at!gpsautostartx=$ena,$mode,$to,$dist,$fr"
    if ! $MCB_TOOLS_DIR/umtscardtool -s at!gpsautostartx=$ena,$mode,$to,$dist,$fr | grep -q OK; then
	syslogger "error" "Modem failed on !gpsautostartx command"
	return 1
    fi
    return 0
}


# AT!GPSFIX=?
# AT!GPSFIX: <fixtype>,<maxtime>,<maxdist>
# <fixtype>: 1-Standalone, 2-MS-Based, 3-MS-Assisted
# <maxtime>: 0-255 seconds
# <maxdist>: 0-4294967280 meters
# AT!GPSFIX=1,15,10
# 
# OK
function gps_gpsfix ()
{
    local mode=1; [ -n "$1" ] && mode=$1
    local to=15; [ -n "$2" ] && to=$2
    local dist=30; [ -n "$3" ] && dist=$3
    syslogger "debug" "Send at!gpsfix=$mode,$to,$dist command "
    if ! $MCB_TOOLS_DIR/umtscardtool -s at!gpsfix=$mode,$to,$dist | grep -q OK; then
	syslogger "error" "Modem failed on !gpsfix command"
	return 1
    fi
    return 0
}

# AT!GPSSTATUS?
# Current time: <y> <m> <d> <dow> <time>
# 
# <y> <m> <d> <dow> <time> Last Fix Status    = <status>
# <y> <m> <d> <dow> <time> Fix Session Status = <status>
# 
# TTFF (sec) = <n>
#
# OK
# <status>:=NONE ACTIVE SUCCESS FAIL
function gps_gpsstatus ()
{
    syslogger "debug" "Send at!gpsstatus? command"
    if ! $MCB_TOOLS_DIR/umtscardtool -s at!gpsstatus?; then
	syslogger "error" "!gpsstatus reports failed"
	return 1
    fi
    return 0
}

# AT!GPSLOC?
# Lat: 50 Deg 34 Min 58.69 Sec N  (0x008FE165)
# Lon: 8 Deg 32 Min 31.62 Sec E  (0x00184C2E)
# Time: 2010 07 07 2 10:58:44 (GPS)
# LocUncAngle: 1  LocUncA: 21  LocUncP: 19  HEPE: f
# 3D Fix
# Altitude: 217  LocUncVe: 20
#
# OK
function gps_gpsloc ()
{
    syslogger "debug" "Send at!gpsloc? command "
    if ! $MCB_TOOLS_DIR/umtscardtool -s at!gpsloc?; then
	syslogger "error" "!gpsloc reports failed"
	return 1
    fi
    return 0
}

# !GPSTRACK: <fixtype>,<maxtime>,<maxdist>,<fixcount>,<fixrate>
# <fixtype>: 1-Standalone, 2-MS-Based, 3-MS-Assisted
# <maxtime>: 0-255 seconds
# <maxdist>: 0-4294967280 meters
# <fixcount>: 1-1000,1000=continuous
# <fixrate>: 1-1,799,999 seconds

# AT!GPSTRACK=1,255,30,1000,1
function gps_gpstrack ()
{
    local mode=1; [ -n "$1" ] && mode=$1
    local to=255; [ -n "$2" ] && to=$2
    local dist=30; [ -n "$3" ] && dist=$3
    local fc=1000; [ -n "$4" ] && fc=$4
    local fr=1; [ -n "$5" ] && fr=$5
    syslogger "debug" "Send at!gpstrack=$mode,$to,$dist,$fc,$fr command "
    if ! $MCB_TOOLS_DIR/umtscardtool -s at!gpstrack=$mode,$to,$dist,$fc,$fr | tr -d '\r'; then
	syslogger "error" "!gpstrack reports failed"
	return 1
    fi
    return 0
}

function gps_gpssatinfo ()
{
    syslogger "debug" "Send at!gpssatinfo? command "
    if ! $MCB_TOOLS_DIR/umtscardtool -s at!gpssatinfo? | tr -d '\r'; then
	syslogger "error" "!gpssatinfo? failed"
	return 1
    fi
    return 0
}

function gps_gpsend ()
{
    syslogger "debug" "Send at!gpsend=0 command "
    if ! $MCB_TOOLS_DIR/umtscardtool -s at!gpsend=0 | tr -d '\r'; then
	syslogger "error" "!gpsend=0 failed"
	return 1
    fi
    return 0
}

#-----------------------------------------------------------------------
# GPS functions
#-----------------------------------------------------------------------

GPS_TIME=
GPS_LASTSTATUS=
GPS_CURRSTATUS=
GPS_LASTERROR=
GPS_CURRERROR=
GPS_TTFF=

function save_gpsstatus ()
{
    gps_gpsstatus | tr -d '\r' | grep -v "OK" >$GPS_STATUS_FILE
}
function save_gpslocation ()
{
    gps_gpsloc | tr -d '\r' | grep -v "OK" >$GPS_LOCATION_FILE
}
function save_gpssatinfo ()
{
    gps_gpssatinfo | tr -d '\r' | grep -v "OK" >$GPS_SATINFO_FILE
}
function get_errorstring ()
{
    if [ -n "$1" ]; then
	echo ${GPS_Errorcode[$1]}
    fi
}
function query_gpsstatus ()
{
    save_gpsstatus
    # cat $GPS_STATUS_FILE
    GPS_TIME=`cat $GPS_STATUS_FILE | sed -n "s/^Current time: \(.*\)$/\1/ p" | tr -d ' :'`
    GPS_TTFF=`cat $GPS_STATUS_FILE | sed -n "s/^TTFF (sec) = \(.*\)$/\1/ p" | tr -d ' :'`
    GPS_LASTSTATUS_TIME=`cat $GPS_STATUS_FILE | sed -n "s/^\(.*\)Last Fix Status    = .*$/\1/p" | tr -d ' :'`
    GPS_CURRSTATUS_TIME=`cat $GPS_STATUS_FILE | sed -n "s/^\(.*\)Fix Session Status = .*$/\1/p" | tr -d ' :'`
    local t1=`cat $GPS_STATUS_FILE | sed -n "s/^.*Last Fix Status    = \(.*\)$/\1/ p"`
    local t2=`cat $GPS_STATUS_FILE | sed -n "s/^.*Fix Session Status = \(.*\)$/\1/ p"`
    GPS_LASTSTATUS=$t1
    GPS_CURRSTATUS=$t2
    GPS_LASTERROR=`echo $t1 | sed -n "s/.*FAILCODE = \(.*\)$/\1/ p"`
    GPS_CURRERROR=`echo $t2 | sed -n "s/.*FAILCODE = \(.*\)$/\1/ p"`
    echo "   GPS time: $GPS_TIME"
    echo "   GPS TTFF: $GPS_TTFF"
    echo "Last status: $GPS_LASTSTATUS_TIME $GPS_LASTSTATUS `get_errorstring $GPS_LASTERROR`"
    echo "Curr status: $GPS_CURRSTATUS_TIME $GPS_CURRSTATUS `get_errorstring $GPS_CURRERROR`"
}
function query_gpsloc ()
{
    save_gpslocation
    cat $GPS_LOCATION_FILE
}
function query_gpssatinfo ()
{
    save_gpssatinfo
    cat $GPS_SATINFO_FILE
}


function issue_gpsfix ()
{
    gps_gpsfix "$1" "$2" "$3"
}
function issue_gpstrack ()
{
    gps_gpstrack "$1" "$2" "$3" "$4" 
}
function wait_gpsfix ()
{
    local rc=0
    while true; do
	query_gpsstatus
	case "$GPS_CURRSTATUS" in
	NONE)
	    syslogger "error" "No gpsfix started"
	    rc=1; break ;;
	ACTIVE)
	    syslogger "debug" "gpsfix started and active"
	    ;;
	SUCCESS)
	    syslogger "debug" "gpsfix succesfully finished"
	    rc=0; break;;
	FAIL,*)
	    syslogger "debug" "gpsfix failed"
	    rc=1; break;;
	*)
	    syslogger "debug" "unknown currstatus $GPS_CURRSTATUS"
	    rc=1; break
	esac

#	case "$GPS_LASTSTATUS" in
#	NONE*)		;;
#	SUCCESS*)	;;
#	FAIL*)		;;
#	*)
#	    syslogger "debug" "unknown laststatus"
#	    rc=1; break
#	esac

	sleep 10
    done

    if [ $rc = 1 ]; then
	syslogger "error" "!gpsstatus reports failed or none fixup"
	#local ec=`echo ${status/\r/} | awk '{print $13}'`
	#echo "Error $ec := ${GPS_Errorcode[$ec]}"
	return 1
    fi
    return 0
}


#-----------------------------------------------------------------------
# GPS Backends
#-----------------------------------------------------------------------

function check_gpsd ()
{
    local pids="`pidof gpsd`"
    [ -z "$pids" ] && return 1
    return 0
}
function start_gpsd ()
{
    query_gpsstatus
    if [ ! "$GPS_CURRSTATUS" = "ACTIVE" ]; then
	syslogger "debug" "restarted GPS tracking for gpsd..."
	issue_gpstrack 1 15 $GPS_TRKPREC
    fi
    
    query_gpsstatus
    if [ "$GPS_CURRSTATUS" = "ACTIVE" ]; then
	if ! check_gpsd; then
	    gpsd -n -G -D 1 -P /var/run/gpsd.pid /dev/ttyUSB2
	else
	    syslogger "debug" "gpsd already running..."
	fi
    else
	syslogger "debug" "Unable to start tracking session"
    fi
}
function stop_gpsd ()
{
    if check_gpsd; then
	killall -TERM gpsd
	rm /var/run/gpsd.pid
    fi
}

function check_ser2net ()
{
    local pids="`pidof ser2net`"
    [ -z "$pids" ] && return 1
    return 0
}
function start_ser2net ()
{
    query_gpsstatus
    if [ ! "$GPS_CURRSTATUS" = "ACTIVE" ]; then
	syslogger "debug" "restarted GPS tracking for ser2net..."
	issue_gpstrack 1 15 $GPS_TRKPREC
    fi
    
    query_gpsstatus
    if [ "$GPS_CURRSTATUS" = "ACTIVE" ]; then
	if ! check_ser2net; then
	    ser2net -C "4000:telnet:0:/dev/ttyUSB2:115200 -LOCAL"
   	else
	    syslogger "debug" "ser2net already running..."
	fi
    else
	syslogger "debug" "Unable to start tracking session"
    fi
}
function stop_ser2net ()
{
    if check_ser2net; then
	killall -TERM ser2net
    fi
}

function check_backend ()
{
    [ $GPS_START_BACKEND -gt ${#GPS_BACKENDS[@]} ] &&
	syslogger "debug" "unknown backend index $GPS_START_BACKEND" &&
	return 1
    case ${GPS_BACKENDS[$GPS_START_BACKEND]} in
    none)	syslogger "debug" "no backend checked" ;;
    gpsd)	check_gpsd ;;
    ser2net)	check_ser2net ;;
    esac
}
function start_backend ()
{
    [ $GPS_START_BACKEND -gt ${#GPS_BACKENDS[@]} ] &&
	syslogger "debug" "unknown backend index $GPS_START_BACKEND" &&
	return 1
    case ${GPS_BACKENDS[$GPS_START_BACKEND]} in
    none)	syslogger "debug" "no backend started" ;;
    gpsd)	start_gpsd ;;
    ser2net)	start_ser2net ;;
    esac
}
function stop_backend ()
{
    [ $GPS_START_BACKEND -gt ${#GPS_BACKENDS[@]} ] &&
	syslogger "debug" "unknown backend index $GPS_START_BACKEND" &&
	return 1
    case ${GPS_BACKENDS[$GPS_START_BACKEND]} in
    none)	syslogger "debug" "no backend stopped" ;;
    gpsd)	stop_gpsd ;;
    ser2net)	stop_ser2net ;;
    esac
}

#-----------------------------------------------------------------------
# MCB GPS Monitor
#-----------------------------------------------------------------------

function gps_start ()
{
    syslogger "debug" "Starting GPS receiver and backend"
    echo "Starting GPS receiver and backend"

    # --- Get an intitial GPS fix ----------------------------
    query_gpsstatus
    if [ "$GPS_CURRSTATUS" = "ACTIVE" ]; then
	syslogger "debug" "Stopping active GPS Fix session"
	gps_gpsend
    fi

    local retries=3
    while [ $retries -gt 0 ]; do
	syslogger "debug" "Trying to obtain initial fix (retries=$retries)"
	issue_gpsfix 1 255 $GPS_FIXPREC
	wait_gpsfix && break
	retries=$[$retries-1]
	query_gpsloc
	query_gpssatinfo
    done

    if [ $retries -eq 0 ]; then
	syslogger "error" "Unable to get GPS fix, yet"
    fi

    start_backend
    touch $GPS_MONITOR_STATUS
}

function gps_stop ()
{
    syslogger "debug" "Stopping GPS receiver and backend"
    echo "Stopping GPS receiver and backend"
    stop_backend
    gps_gpsend
    rm $GPS_MONITOR_STATUS
}

function gps_monitor ()
{
    if [ "$GPS_ENABLE" -eq "1" ]; then
	if [ ! -e $GPS_MONITOR_STATUS ]; then
	    syslogger "debug" "Starting GPS receiver"
	    gps_start
	    
	fi
	if ! check_backend; then
	    syslogger "debug" "Starting GPS backend (again)"
	    start_backend
	fi
	query_gpsstatus
	query_gpsloc
	query_gpssatinfo
    else
	syslogger "debug" "GPS monitoring diabled"
    fi
}

function print_usage {
    echo "Usage: -h <start|stop|monitor>"
}

echo -e "\nkonzeptpark GPS monitor\n"

while getopts "hfd:s:r:" optionName; do
	case "$optionName" in
	h) print_usage; exit 1;;
#	f) force=1;;			# FIXME: Set values for to,dist, ...
#	d) device="$OPTARG";;
#	s) serverip="$OPTARG";;
#	r) versionpath="$OPTARG";;
	[?]) print_usage; exit 1;;
	esac
done

if ! [ $# -ge 1 ]; then
    print_usage; exit 1
fi

obtainlock $GPS_MONITOR_PID_FILE
syslogger "debug" "Started monitor (`date`)"

# 'DetectModemCard' must becalled before, so MODEM_STATUS might be empty
ReadModemStatusFile
if [ -z "$MODEM_STATUS" -o "x$MODEM_STATUS" == "x${MODEM_STATES[no_modemID]}" ]; then
    DetectModemCard
fi

if gps_checkmodem_feature; then

    case "$1" in
	"start")	gps_start ;;
	"stop")	gps_stop ;;
	"monitor")	gps_monitor ;;
    esac

else
    echo "Modem doesn't support GPS"
fi

syslogger "debug" "Finished monitor (`date`)"
releaselock
